### DISCUSSION TITLE
Transform std::expected into error or value

### DISCUSSION THREAD

**Discussion by mathisloge** [NVIDIA/stdexec#1433] [Q&A] [answered]
Is it somehow possible to transform a result of type std::expected into either an error state or a valid value? 
E.g. 

```cpp
stdexec::schedule(scheduler) | stdexec::then([](){
   std::expected<int, MyError> my_operation = do_some_work();
   if(my_operation.has_value()) { return *my_operation; };
   
   // now set an error for MyError
}) 
| stdexec::upon_error([](MyError error) { // handleError })
| stdexec::then([](int value){});
...

```
 
Hope you know what I'm trying to do.

**Comment by maikel** [comment]
In principle that is doable but not intrinsically with then. You can define your own adaptor (which is more optimized) either by hand or by using let_value in conjunction with variant_sender, just, and just_error (which might be easier to do).

**Reply by mathisloge** [reply]
Do you happen to know if there are some simple examples to custom adaptors? 
E.g. is it always the pattern as in https://github.com/NVIDIA/stdexec/blob/main/examples/algorithms/then.hpp 
Have a `_<op>_sender` and then a  `_<op>_receiver` which either calls `stdexec::set_value` or `stdexec::set_error`?

**Reply by maikel** [reply, marked as answer]
The implementation in stdexec optimizes for various things. 

I made this example not very long ago for a naive implementation of the then sender, maybe that helps you. https://godbolt.org/z/5qczb9dhh

I will try to make a tutorial on sender adaptors soonish, but maybe someone else has already done so.

**Reply by mathisloge** [reply]
Thank you
